admin/services/attendance.service.js

```diff
@@ -64,6 +64,24 @@ const attendanceService = {
    */
   getAttendanceCodes: async () => {
     return attendanceCodeDataService.sqlFindAttendanceCodes()
+  },
+
+  /**
+   * Check if pupil has an existing attendance entry for the
+   * specific `pinEnv` environment.
+   *
+   * All attendance entries are considered for live checks, while for
+   * familiarisation checks, only `left school` entries are considered
+   *
+   * @param pupilId
+   * @param pinEnv
+   * @return {boolean}
+   */
+  hasAttendance: async (pupilId, pinEnv) => {
+    const pupilAttendance = await pupilAttendanceDataService.findOneByPupilId(pupilId)
```

Comment:
Which roles would be performing this action?

Reply:
If you mean user roles, this will be called when teachers get the list of pupils to generate pins for.

Reply:
That’s not directly called by the teacher (by the controller), it’s used by `pinGenerationService.isValid` which is, as you mention, called by `pinGenerationService.getPupils(dfeNumber, pinEnv)` with `dfeNumber`.

The idea is that `getPupils` will get the pupils list using the `dfeNumber`, then will check if the pupils are valid using `isValid` (which also calls `hasAttendance` for a specific pupil) — this functionality was previously in `isValid`. I thought it makes more sense to put it in that separate function to keep the `isValid` one cleaner, and write unit tests for `hasAttendance` in particular 😄

Reply:
That seems reasonable to me.  👍